### PR TITLE
*: configure http client from config

### LINKED ISSUE #853: What's the correct way to config Alertmanager to use a forward proxy?
In this [thread](https://github.com/coreos/prometheus-operator/issues/399#issuecomment-307114123) with @brancz, I was seeking a way to config Alertmanager to use a forward proxy. One way is to use the `https_proxy` env var detailed [here](https://github.com/prometheus/alertmanager/issues/628).

Is this the correct way to do it in the long haul, or should it be in Alertmanager's config. Thanks!

### PR DESCRIPTION
**Pull request by brancz**
It seems that many people are interested in configuring the HTTP client used to perform requests to the notification providers. As @Conorbro recently added support for this in the common/config package, I made use of it here.

Closes #853

@stuartnelson3 @fabxc @mxinden @Gouthamve @brian-brazil

### CODE REVIEW

File: config/notifiers.go [hunk 1]
```diff
@@ -176,6 +178,8 @@ func (c *EmailConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
 type PagerdutyConfig struct {
 	NotifierConfig `yaml:",inline" json:",inline"`
 
+	HTTPConfig *commoncfg.HTTPClientConfig `yaml:"http_config,omitempty" json:"http_config,omitempty"`
```
Comment (brian-brazil) [inline config/notifiers.go]: Should this go in NotifierConfig considering we use it everywhere?

File: notify/impl.go [hunk 2]
```diff
@@ -141,13 +142,13 @@ var userAgentHeader = fmt.Sprintf("Alertmanager/%s", version.Version)
 // Webhook implements a Notifier for generic webhooks.
 type Webhook struct {
 	// The URL to which notifications are sent.
```
Comment (brian-brazil) [inline notify/impl.go]: Comment needs an update

Comment (brancz) [inline notify/impl.go]: Not sure why the URL was exported before, I moved the `Config` -> `conf` to be consistent with all the other implementations.

File: config/notifiers.go (hunk 1 above)
Comment (brancz) [inline config/notifiers.go]: We have it everywhere except email configs, it would be odd to be able to have it there, no?

Comment (brian-brazil) [inline config/notifiers.go]: That's true.

**Comment by brancz** [comment]
If there are no general concerns anymore I'll start working on docs.

**Comment by brancz** [comment]
I just talked to @fabxc, he has some comments on the general topic of the common package. I'll hold off on merging util he has a look.

**Comment by fabxc** [comment]
I've plenty of concerns on this – but that's a longer discussion I hope to find time for soonish.

Where else is this struct being used outside of AM?

**Comment by brian-brazil** [comment]
Right now the blackbox exporter. We likely also need it in the consul exporter.

**Comment by brancz** [comment]
@fabxc @brian-brazil how should we continue on this, as it should be possible for people to configure this somehow.

**Comment by brian-brazil** [comment]
I've been proposing code in common for this for 18 months now, which is what the blackbox exporter uses. @fabxc is against us having common packages for this, while I strongly believe code like this (particularly security related) must come from a common library.

**Comment by brancz** [comment]
> I've plenty of concerns on this – but that's a longer discussion I hope to find time for soonish.

I guess this would be the start of having this discussion, so we can come to an agreement of how to proceed.

**Comment by simonpasquier** [comment]
To set the discussion, here is a recap of the concrete use cases that call for more flexibility on the HTTP client configuration:

* Use bearer tokens for web-hook services running behind proxies like [kube-rbac-proxy](https://github.com/brancz/kube-rbac-proxy) and [oauth-proxy](https://github.com/openshift/oauth-proxy).
* Override TLS parameter (skip certificate verification, custom CA, ...).
* Use HTTP proxies.

**Comment by brian-brazil** [comment]
We're all good with the use cases, it's exactly how we share the common code that there is disagreement on.

**Comment by ddewar2k** [comment]
Moving my question to this more recent thread. I still don't see code that supports http proxy for receivers and it looks like all the threads on this closed. Is my understanding correct?

**Comment by simonpasquier** [comment]
@brancz @brian-brazil @fabxc Now that Prometheus itself uses http client from the common library  (see https://github.com/prometheus/prometheus/pull/3674), can we reconsider this PR?
FWIW I've rebased the PR on the master branch in my local fork and after resolving the conflicts, it seems to be working fine.

**Comment by brancz** [comment]
I'm happy with going forward with this.

**Comment by simonpasquier** [comment]
@brancz, cool! This is your PR rebased on master: https://github.com/prometheus/alertmanager/compare/master...simonpasquier:httpcfg
If you don"t have time, I can carry on the work.

**Comment by brancz** [comment]
@simonpasquier I don't know when I would get to it, so if you can, go for it!

**Comment by simonpasquier** [comment]
@brancz I've created #1222.

**Pull request closed without merging**